File: examples/python/dex_examples/products/polling/polling_flow.py
# Copyright (c) 2022-2026 Super Durable, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from datetime import timedelta

from dex import (
    Attribute,
    Channel,
    Context,
    Flow,
    PersistenceSchema,
    Step,
    StepDecision,
    StepList,
    StepMovement,
    Timer,
    Wait,
    dead_end,
    go_to,
    go_to_multi,
    graceful_complete,
)

from dex_examples.shared.my_dependency_service import MyDependencyService

TASK_A_COMPLETED = "task-a-completed"
TASK_B_COMPLETED = "task-b-completed"
TASK_C_COMPLETED = "task-c-completed"


class WaitForTasks(Step[None]):
    def __init__(
        self,
        task_a_completed: Channel[None],
        task_b_completed: Channel[None],
        task_c_completed: Channel[None],
    ) -> None:
        self.task_a_completed = task_a_completed
        self.task_b_completed = task_b_completed
        self.task_c_completed = task_c_completed

    def wait_for(self, context: Context, input: None) -> Wait:
        del context, input
        return Wait.all_of(
            self.task_a_completed.for_one(),
            self.task_b_completed.for_one(),
            self.task_c_completed.for_one(),
        )

    def execute(self, context: Context, input: None) -> StepDecision:
        del context, input
        return graceful_complete("all tasks completed")


class Poll(Step[int]):
    def __init__(
        self,
        service: MyDependencyService,
        current_polls: Attribute[int],
        task_c_completed: Channel[None],
    ) -> None:
        self.service = service
        self.current_polls = current_polls
        self.task_c_completed = task_c_completed

    def wait_for(self, context: Context, input: int) -> Wait:
        del context, input
        return Wait.until(Timer.by_duration(timedelta(seconds=1)))

    def execute(self, context: Context, input: int) -> StepDecision:
        self.service.call_api1("calling API1 for polling service C")
        polls = self.current_polls.get(context)
        if polls >= input:
            self.task_c_completed.publish(context, None)
            return dead_end()
        self.current_polls.set(context, polls + 1)
        return go_to(self, input)


class Initialize(Step[int]):
    def __init__(
        self,
        current_polls: Attribute[int],
        poll: Poll,
        wait_for_tasks: WaitForTasks,
    ) -> None:
        self.current_polls = current_polls
        self.poll = poll
        self.wait_for_tasks = wait_for_tasks

    def execute(self, context: Context, input: int) -> StepDecision:
        self.current_polls.set(context, 0)
        return go_to_multi(
            StepMovement.of(self.poll, input),
            StepMovement.of(self.wait_for_tasks, None),
        )


class PollingFlow(Flow[int]):
    current_polls = Attribute("current-polls", int)
    task_a_completed = Channel[None](TASK_A_COMPLETED, type(None))
    task_b_completed = Channel[None](TASK_B_COMPLETED, type(None))
    task_c_completed = Channel[None](TASK_C_COMPLETED, type(None))

    def __init__(self, service: MyDependencyService) -> None:
        self.service = service
        self.poll = Poll(service, self.current_polls, self.task_c_completed)
        self.wait_for_tasks = WaitForTasks(
            self.task_a_completed,
            self.task_b_completed,
            self.task_c_completed,
        )
        self.initialize = Initialize(
            self.current_polls,
            self.poll,
            self.wait_for_tasks,
        )

    def get_steps(self) -> StepList[int]:
        return StepList.start_step(self.initialize).other_steps(
            self.poll,
            self.wait_for_tasks,
        )

    def get_persistence_schema(self) -> PersistenceSchema:
        return PersistenceSchema.of(
            self.current_polls,
            self.task_a_completed,
            self.task_b_completed,
            self.task_c_completed,
        )

File: examples/typescript/src/products/polling/polling-flow.ts
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Attribute,
  Channel,
  StepList,
  StepMovement,
  Timer,
  Wait,
  deadEnd,
  doubleCodec,
  goTo,
  goToMany,
  gracefulComplete,
  voidCodec,
  type Context,
  type Flow,
  type PersistenceSchema,
  type Step,
  type StepDecision,
} from "@superdurable/dex";

import {
  myDependencyService,
  type MyDependencyService,
} from "../../shared/my-dependency-service.js";

export const TASK_A_COMPLETED = "task-a-completed";
export const TASK_B_COMPLETED = "task-b-completed";
export const TASK_C_COMPLETED = "task-c-completed";

const POLL_INTERVAL_MS = 1_000;

export const taskACompleted = new Channel(TASK_A_COMPLETED, voidCodec);
export const taskBCompleted = new Channel(TASK_B_COMPLETED, voidCodec);
export const taskCCompleted = new Channel(TASK_C_COMPLETED, voidCodec);

export class PollingFlow implements Flow<number> {
  public readonly currentPolls = new Attribute("current-polls", doubleCodec);

  public readonly initialize = new Initialize(this);
  public readonly poll = new Poll(this);
  public readonly waitForTasks = new WaitForTasks(this);

  public constructor(public readonly service: MyDependencyService = myDependencyService) {}

  public getFlowType(): string {
    return "PollingFlow";
  }

  public getSteps() {
    return StepList.startStep(this.initialize).otherSteps(this.poll, this.waitForTasks);
  }

  public getPersistenceSchema(): PersistenceSchema {
    return {
      attributes: [this.currentPolls],
      channels: [taskACompleted, taskBCompleted, taskCCompleted],
    };
  }
}

class Initialize implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public constructor(private readonly flow: PollingFlow) {}

  public getStepType(): string {
    return "Initialize";
  }

  public execute(context: Context, maximumPolls: number): StepDecision {
    this.flow.currentPolls.set(context, 0);
    return goToMany(
      StepMovement.of(Poll, maximumPolls),
      StepMovement.of(WaitForTasks, undefined),
    );
  }
}

class WaitForTasks implements Step<void> {
  public constructor(private readonly flow: PollingFlow) {}

  public getStepType(): string {
    return "WaitForTasks";
  }

  public waitFor(_context: Context, _input: void): Wait {
    return Wait.allOf(
      taskACompleted.forOne(),
      taskBCompleted.forOne(),
      taskCCompleted.forOne(),
    );
  }

  public execute(_context: Context, _input: void): StepDecision {
    return gracefulComplete("all tasks completed");
  }
}

class Poll implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public constructor(private readonly flow: PollingFlow) {}

  public getStepType(): string {
    return "Poll";
  }

  public waitFor(_context: Context, _maximumPolls: number): Wait {
    return Wait.until(Timer.byDuration(POLL_INTERVAL_MS));
  }

  public execute(context: Context, maximumPolls: number): StepDecision {
    this.flow.service.callAPI1("calling API1 for polling service C");
    const polls = this.flow.currentPolls.get(context);
    if (polls >= maximumPolls) {
      taskCCompleted.publish(context, undefined);
      return deadEnd();
    }
    this.flow.currentPolls.set(context, polls + 1);
    return goTo(Poll, maximumPolls);
  }
}

export const pollingFlow = new PollingFlow();
